packages/@aws-cdk/aws-apprunner/lib/service.ts

```diff
@@ -860,71 +1030,134 @@ export class Service extends cdk.Resource {
     this.serviceStatus = resource.attrStatus;
     this.serviceName = resource.ref;
   }
+
+  /**
+   * This method adds an environment variable to the App Runner service.
+   */
+  public addEnvironment(name: string, value: string) {
+    this.variables.push({ name: name, value: value });
+  }
+
+  /**
+   * This method adds a secret as environment variable to the App Runner service.
+   */
+  public addSecret(name: string, secret: Secret) {
+    if (!this.instanceRole) {
+      this.instanceRole = this.createInstanceRole();
+    }
+    secret.grantRead(this.instanceRole);
+    this.secrets.push({ name: name, value: secret.arn });
+  }
+
+  /**
+   * This method generates an Instance Role. Needed if using secrets and props.instanceRole is undefined
+   * @returns iam.IRole
+   */
+  private createInstanceRole(): iam.IRole {
+    return new iam.Role(this, 'InstanceRole', {
+      assumedBy: new iam.ServicePrincipal('tasks.apprunner.amazonaws.com'),
+      roleName: cdk.PhysicalName.GENERATE_IF_NEEDED,
+    });
+  }
+
+  /**
+   * This method generates an Access Role only when ImageRepositoryType is ECR and props.accessRole is undefined
+   * @returns iam.IRole
+   */
+  private generateDefaultRole(): iam.Role {
+    const accessRole = new iam.Role(this, 'AccessRole', {
+      assumedBy: new iam.ServicePrincipal('build.apprunner.amazonaws.com'),
+    });
+    accessRole.addToPrincipalPolicy(new iam.PolicyStatement({
+      actions: ['ecr:GetAuthorizationToken'],
+      resources: ['*'],
+    }));
+    this.accessRole = accessRole;
+    return accessRole;
+  }
+
+  private getEnvironmentSecrets(): { [key: string]: Secret } | undefined {
+    return this.source.codeRepository?.codeConfiguration.configurationValues?.environmentSecrets ||
+      this.source.imageRepository?.imageConfiguration?.environmentSecrets;
```

Comment:
They are meant to be mutex, either you create a service with an image from ECR or you build it from a code repo. What I can see, is that there's no logic implemented at the API lvl blocking this behavior, and it will be done from the CFN instead. Maybe an exception should be made right in the constructor. 

Something like:

`if (this.source.imageRepository && this.source.codeRepository) {
  throw new Error('You can only supply either \'imageRepository\' and \'codeRepository\' at a time.');
}`

Reply:
Actually it's mutex because the SourceConfig and the Source definition. There's no way both imageRepository and codeRepository are in the same invocation.